src/nix-build/nix-build.cc

```diff
@@ -0,0 +1,482 @@
+
+using namespace nix;
+using std::stringstream;
+
+extern char ** environ;
+
+std::vector<string> shellwords(const string & s)
+{
+    auto whitespace = std::regex("^(\\s+).*");
+    auto begin = s.cbegin();
+    auto res = std::vector<string>{};
+    auto cur = stringstream{};
+    enum state {
+        sBegin,
+        sQuote
+    };
+    state st = sBegin;
+    auto it = begin;
+    for (; it != s.cend(); ++it) {
+        if (st == sBegin) {
+            auto match = std::smatch{};
+            if (regex_search(it, s.cend(), match, whitespace)) {
+                cur << string(begin, it);
+                res.push_back(cur.str());
+                cur = stringstream{};
+                it = match[1].second;
+                begin = it;
+            }
+        }
+        switch (*it) {
+            case '"':
+                cur << string(begin, it);
+                begin = it + 1;
+                st = st == sBegin ? sQuote : sBegin;
+                break;
+            case '\\':
+                /* perl shellwords mostly just treats the next char as part of the string with no special processing */
+                cur << string(begin, it);
+                begin = ++it;
+                break;
+        }
+    }
+    cur << string(begin, it);
+    auto last = cur.str();
+    if (!last.empty()) {
+        res.push_back(std::move(last));
+    }
+    return res;
+}
+
+int main(int argc, char ** argv)
+{
+    return handleExceptions(argv[0], [&]() {
+        initNix();
+        auto store = openStore();
+        auto dryRun = false;
+        auto verbose = false;
+        auto runEnv = std::regex_search(argv[0], std::regex("nix-shell$"));
+        auto pure = false;
+        auto fromArgs = false;
+        auto packages = false;
+        auto interactive = true;
+
+        auto instArgs = Strings{};
+        auto buildArgs = Strings{};
+        auto exprs = Strings{};
+
+        auto shell = getEnv("SHELL", "/bin/sh");
+        auto envCommand = string{}; // interactive shell
+        auto envExclude = Strings{};
+
+        auto myName = runEnv ? "nix-shell" : "nix-build";
+
+        auto inShebang = false;
+        auto script = string{};
+        auto savedArgs = std::vector<string>{};
+
+        auto tmpDir = AutoDelete{createTempDir("", myName)};
+
+        auto outLink = string("./result");
+        auto drvLink = (Path) tmpDir + "/derivation";
+
+        auto args = std::vector<string>{};
+        for (int i = 1; i < argc; ++i)
+            args.push_back(argv[i]);
+        // Heuristic to see if we're invoked as a shebang script, namely, if we
+        // have a single argument, it's the name of an executable file, and it
+        // starts with "#!".
+        if (runEnv && argc > 1 && !std::regex_search(argv[1], std::regex("nix-shell"))) {
+            script = argv[1];
+            if (access(script.c_str(), F_OK) == 0 && access(script.c_str(), X_OK) == 0) {
+                auto SCRIPT = std::ifstream(script);
```

Comment:
Also I would use `tokenizeString(readFile(script), "\n")` to read the file in lines.